Lvmdbusd and lvmdbustest bug fixes and coverage improvements

tasleson opened the pull request.

tasleson committed:
lvmdbustest: Use updated pyudev syntax

Previous syntax has been deprecated.

tasleson committed:
lvmdbustest: Include major number 259

When you have > 16 partitions for a block device the major number
changes, include them for testing.

tasleson committed:
lvmdbusd: Remove addl. checks on --nojson

We check earlier that if you specify --nojson we will exit.

tasleson committed:
lvmdbusd: Move arg. setup/checking to function

tasleson committed:
lvmdbusd: Add lockfile

The daemon cannot handle multiple copies of itself running at the
same time, ensure this cannot happen.

tasleson committed:
lvmdbustest: Remove force exception in _wait_for_job

We put this in to test one of the paths in the damon, but unfortunately
if we hit the race condition where the job actually is done we will try
to call j.Wait(1) after the remove.  This fails with:

dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod:
Method "Wait" with signature "i" on interface "com.redhat.lvmdbus1.Job"
doesn't exist

Which is caused by the dbus object no longer existing.  We could handle
this, but the issue is we no longer have the ability to get the result to
return, they have been lost.

A better solution would be to write a specific unit test to force this code
path and handle all the possible outcomes.

tasleson committed:
lvmdbusd: Re-work flight recorder data

Introduce a new lock for the flight recorder, so that we can dump it when
a command is block waiting for lvm to complete.  Also in all paths we will
addthe metadata to the flight recorder before it's done, so we will have
it when a command hangs and we dump the flight recorder. Add the missing
bits after the command has finished.

Cleaned up the output too.

tasleson committed:
lvmdbusd: Add SIGUSR2 to dump flight recorder

tasleson committed:
lvmdbusd: Remove unneeded command line args

tasleson committed:
lvmdbusd: Remove exclusionary language

tasleson committed:
lvmdbustest: nesting improvements

tasleson committed:
lvmdbusd: Handle no lastlog

Depending on when an occurs, it maynot have any information available for
lastlog.  In this case try to grab an error message from the original
response.

tasleson committed:
lvmdbusd: Correct typos

tasleson committed:
lvmdbusd: Make sure to set cfg.got_external_event

We were incorrectly only setting this if --udev wasn't present on the
command line.  In all cases when we see a manager.ExternalEvent we want
to set this.

tasleson committed:
lvmdbusd: Remove unused locking functionality

I don't think we have ever utilized this, remove.

tasleson committed:
lvmdbustest: Add test for passing log file in options

tasleson committed:
lvmdbustest: Add test for external event

Ensure that when we trigger an external event that we don't incorrectly
handle multiple --config options.

tasleson committed:
lvmdbustest: Add test removing incomplete job

When you try to remove a job that is incomplete you get a dbus exception.
Test for this error condition.

tasleson committed:
lvmdbusd: fix hangs on SIGINT

Rather than trying to bubble up return codes that get us to exit cleanly
it's better to just raise an exception to bail.  In some cases functions
don't have return codes, so they cannot be checked.

tasleson committed:
lvmdbusd: Add debug circular buffer

When the daemon isn't started with --debug we will keep a circular
buffer of the past N number of debug messages which we will output
when we encounter an issue.

tasleson committed:
lvmdbusd: Correct typos

tasleson committed:
lvmdbustest: Add DaemonInfo class

This class handles identifying daemon, sending signals to it, and starting
it back up again.

tasleson committed:
lvmdbustest: Factor out tearDown implementation for re-use

tasleson committed:
lvmdbustest: Add optional option to _create_num_lvs

This allows us to create the LVs async., thus queuing them up in daemon.

tasleson committed:
lvmdbustest: Add test_sigint test

Get the daemon busy and send it SIGINT to ensure the daemon exits.

tasleson committed:
lvmdbustest: Add test for ensuring only 1 instance of daemon

tasleson committed:
lvmdbustest: Add systemctl daemon start

tasleson committed:
lvmdbusd: Set LVM_COMMAND_PROFILE=lvmdbusd

We need this to prevent lvm from interleaving the JSON output with errors
written to stderr.

tasleson committed:
lvmdbustest: Skip test_singleton_daemon running systemd svc.

tasleson commented:
@vojtechtrefny please take a peek at these changes.  I didn't add always gathering debug data to a file, yet :smiley:

tasleson committed:
lvmdbusd: WS, imports, grammar

tasleson committed:
lvmdbusd: Don't report recoverable error

Lvm occasionally fails to return all the request JSON keys in the output of
"fullreport".  This happens very rarely.  When it does the daemon was reporting
the resulting informational exception:

MThreadRunner: exception
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/lvmdbusd/utils.py", line 667, in _run
    self.rc = self.f(*self.args)
  File "/usr/lib/python3.9/site-packages/lvmdbusd/fetch.py", line 40, in _main_thread_load
    (lv_changes, remove) = load_lvs(
  File "/usr/lib/python3.9/site-packages/lvmdbusd/lv.py", line 143, in load_lvs
    return common(
  File "/usr/lib/python3.9/site-packages/lvmdbusd/loader.py", line 37, in common
    objects = retrieve(search_keys, cache_refresh=False)
  File "/usr/lib/python3.9/site-packages/lvmdbusd/lv.py", line 95, in lvs_state_retrieve
    l['vdo_operating_mode'],
KeyError: 'vdo_operating_mode'

The daemon retries the operation, which usually works and the daemon continues.
However, simply reporting this informational stack trace is causing CI and other
automated tests to fail as they expect no tracebacks in the log output.

Remove the reporting of this code path unless it persists and causes the daemon
to give up and exit.

Ref: https://bugzilla.redhat.com/show_bug.cgi?id=2120267

tasleson committed:
lvmdbusd: Handle SIGINT quietly

Change how we exit on SIGINT so that we don't output needless debug.

tasleson committed:
lvmdbusd: Re-work error handling for run_cmd

Instead of lumping all the exceptions, break them out to handle the dbus
exceptions separately, to reduce the amount of debug information that ends
up in the journal that has questionable value.

tasleson committed:
lvmdbusd: refactor and correct fetch thread logic

Simplify the fetch thread and correct the logic used for selecting the
options which are used when we batch update a state refresh.

vojtechtrefny commented on test/dbus/lvmdbustest.py:
```diff
@@ -465,6 +463,11 @@ def tearDown(self):
+	def tearDown(self):
+		# If we get here it means we passed setUp, so lets remove anything
+		# and everything that remains, besides the PVs themselves
+		self.clean_up()
```
[addCleanup](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup) might be useful here, we use it in our tests because `tearDown` doesn't run if `setUp`.

tasleson replied:
Thanks, added

vojtechtrefny commented on daemons/lvmdbusd/manager.py:
```diff
@@ -266,3 +266,12 @@ def Uuid(self):
+	@dbus.service.method(
+		dbus_interface=MANAGER_INTERFACE,
+		in_signature='a{sv}')
+	def SetCmdLineOptions(self, options):
```
I am little bit afraid this could cause a lot of issues when two different tools both use this function to set their own custom command line options to add.

tasleson replied:
> I am little bit afraid this could cause a lot of issues when two different tools both use this function to set their own custom command line options to add.

Indeed.  With the current implementation, the options replace whatever was present before.  So multiple clients would step on each other.  I'll remove this change from patch set.

vojtechtrefny reviewed (approved):
Looks good to me in general, but I have some concerns about the new `SetCmdLineOptions` function.

tasleson committed:
lvmdbustest: Skip test_nesting if scan_lvs is not enabled

Commit:
tests: fsadm 300m xfs

Commit:
tests: remove slash

Commit:
tests: check for at least 1G of RAM

This test requires quite some RAM to be there.

Commit:
tests: make damage optional

Investigate how to damage thin-pool across different versions/platforms.
Until them - use  'should' for test result.

Commit:
tests: no caching for blkid in testing

Commit:
tests: lvextend update for 300M XFS

tasleson committed:
lvmdbusd: Don't output debug for fullreport exit code 5

This is expected with an exported vg

tasleson committed:
lvmdbusd: Use common function for traceback

We were using a number of different ways to achieve the same result.  Use
a common function to make this consistent.

tasleson committed:
lvmdbusd: Re-work error handling

Introduce an exception which is used for known existing issues with lvm.
This is used to distinguish between errors between lvm itself and lvmdbusd.
In the case of lvm bugs, when we simply retry the operation we will log
very little.  Otherwise, we will dump a full traceback for investigation
when we do the retry.

tasleson committed:
lvmdbusd: Instruct lvm to output debug to file for fullreport

Historically we have seen a few different errors which occur when we call
fullreport.  Failing exit code and JSON which is missing one or more keys.
Instruct lvm to dump the debug to a file during fullreport calls when we
fork & exec lvm. If we encounter an error, ouput the debug data.
The reason this isn't being done when lvmshell is used is because we
don't have an easy way to test the error paths.

This change is complicated by the following:

1. We don't know if fullreport was good until we evaluate all the JSON.
   This is done a bit after we have called into lvm and returned.
2. We don't want to orphan the debug file used by lvm if the daemon is
   killed. Thus we try to minimize the window where the debug file hasn't
   already been unlinked.  A RFE to pass an open FD to lvm for this
   purpose is outstanding.

The temp. file is:
-rw------. 1 root root /tmp/lvmdbusd.lvm.debug.XXXXXXXX.log

tasleson committed:
lvmdbusd: Raise LvmBug exception for invalid JSON

This will cause lvm debug data to get logged if it's available.

tasleson committed:
lvmdbusd: Remove duplicate code

The logic for _cache_lv and _writecache_lv was identical except for which
underlying lvm command to run.  Factor out common.

tasleson committed:
lvmdbustest: Utilize addCleanup in unit test

Register the clean up with addCleanup so we ensure clean_up gets run
regardless of what happens in setUp.

tasleson committed:
lvmdbusd: Remove PID from log messages

Previously the daemon would output PID:TID.  If it's running under systemd
it skips outputting PID as systemd already does this.

tasleson commented:
@vojtechtrefny This PR for some reason has commits from others even though everything looks OK in my repo. after I updated and rebased, thinking something amiss at github.  I duplicated the branch and created a new PR: https://github.com/tasleson/lvm2/pull/16 and it looks as expected.

Changes in this PR (and the identical new PR) includes
* Removed the support for global config setting changes
* Added use of `addCleanup` in unit test
* Added collecting lvm debug for `fullreport`.  This won't interfere with anaconda as end users don't have access to the fullreport run within the daemon to update state.
* Re-worked error handling, introduce new exception `LvmBug`
* Skip `test_nesting` unit test if scan_lvs isn't set
* Some other misc. fixes, like remove some duplicate code.

If you could look at the commit bf5bf4464 `lvmdbustest: Skip test_nesting if scan_lvs is not enabled` and newer that would be great, thanks!

Commit:
tests: rework makefile

Improve dependency tracking
Use 'foreach()' with better tracing support
Link whole 'shell' dir instead of linking individual *.sh files

Commit:
gitignore: update

Commit:
make: update make.tmpl

Add new define 'newline' for use in 'foreach()'
Add new $(SHOW) for makefile printing output
Add 'make print-VAR' for easier debugging of Makefiles' variables.

Commit:
gcc: eliminate warnings

Gcc starts to show new warning - although unlikely to be able to hit
initialize variables to 0.

prajnoha committed:
report: fix lv_active column type from STR to BIN

Fix lv_active to be of BIN type instead of STR. This allows lv_active to
follow the report/binary_values_as_numeric setting as well as --binary
cmd line switch. Also, it makes it possible to use -S|--select with
either textual or numeric representation of the value, like 'lvs -S
active=active' but also 'lvs -S active=1'.

Commit:
python: variable without destdir

PYTHON_PREFIX without $(DESTDIR)

Commit:
tests: vdo emulation without vdo binary

Avoid inserting 'vdo' binary into path - and use
alias and VDO_BINARY shell vars for emulation.

Commit:
tests: add thin requirement

Commit:
tests: add /dev requirement

This test could only be run when user passes  LVM_TEST_DEVDIR=/dev
as it requires and expects actions to be going in this dir, skip
otherwise.

Also 'extend_filter' manages multiple args in on lvm.conf update.

Commit:
tests: use conv=fdatasync

When we only need wait for result - avoid using slow 'oflag=sync'
with virtual layers

Commit:
tests: writecache adapt to 300M XFS

tasleson committed:
lvmdbusd: Use pseudo tty to get "lvm>" prompt again

When lvm is compiled with editline, if the file descriptors don't look like
a tty, then no "lvm> " prompt is done.  Having lvm output the shell prompt
when consuming JSON on a report file descriptor is very useful in
determining if lvm command is complete.

tasleson committed:
lvmdbusd: Handle 'exit' for lvm_shell_proxy

Useful for testing `exit_shell` when running interactively.

Commit:
tests: cache use 300M XFS

Commit:
tests: skip this test for 5.19

mdraid has some breakage - so 5.19 is crashing
(possibly even some more older version - that can be added as well)
Test seems to pass with 6.0-rc kernel.

tasleson committed:
lvmdbustest: Move signals to last

When we do the signal testing we default back to fork & exec.  Put these
on the end to maximize our lvm shell testing until we fix this proper.

tasleson committed:
lvmdbustest: Check for self.pvs

If we don't make it through setUp, self.pvs will not exist.

tasleson committed:
lvmdbusd: Correct undefined var

tasleson committed:
lvmdbusd: Use common func. for checking missing LV keys

tasleson committed:
lvmdbusd: Check for KeyError in refresh

Bubble up a LvmBug if we get a KeyError on a lvm column name.

tasleson committed:
lvmdbusd: Correct log_error for shell proxy

tasleson committed:
lvmdbusd: Raise IOError instead of ignoring

We end up in a bad state if we simply eat IOErrors here.  Exit the lvmshell
process and raise the IOError.

tasleson committed:
lvmdbusd: Call readlines instead of readline

Better to drain everything we have now that our IO is line orientated
when using a ptty.

tasleson committed:
lvmdbusd: Include lvm debug output for lvmshell

Move the option to add the debug file into lvm_full_report_json so that
we collect the debug data when we fork & exec lvm and when we use lvm
shell.

teigland committed:
fix fs block size detection

blkid_get_tag_value() is not a reliable way to query the fs
BLOCK_SIZE, so use the "probe" functions instead.

tasleson committed:
lvmdbustest: Add wrapper for injecting errors

Useful for forcing lvmdbusd down error paths.
